src/libcore/ops/drop.rs

```diff
@@ -18,63 +18,9 @@
 ///
 /// ## Implementing `Drop`
 ///
-/// When a value goes out of scope, the compiler automatically insert a call to `Drop::drop` on that value. For example,
-///
-/// ```rust
-/// struct HasDrop;
-///
-/// impl Drop for HasDrop {
-///     fn drop(&mut self) {
-///         println!("Dropping!");
-///     }
-/// }
-///
-/// fn main() {
-///     let _x = HasDrop;
-/// }
-/// ```
-///
-/// Here, `main` will print `Dropping!`. In other words, the compiler generates code that
-/// kind of looks like this:
-///
-/// ```rust,compile_fail,E0040
-/// # struct HasDrop;
-/// #
-/// # impl Drop for HasDrop {
-/// #     fn drop(&mut self) {
-/// #         println!("Dropping!");
-/// #     }
-/// # }
-/// fn main() {
-///     let _x = HasDrop;
-///
-///     let mut x = _x;
-///     Drop::drop(&mut x);
-/// }
-/// ```
-///
-/// Our custom implementation of `Drop` will be called at the end of `main`.
-///
-/// ## You cannot call `Drop::drop` yourself
-///
-/// Because the compiler automatically calls `Drop::drop`, you cannot call it yourself. This
-/// would lead to a "double drop", where `Drop::drop` is called twice on the same value. This
-/// can lead to things like "double frees".
-///
-/// In other words, if you tried to write the code in the previous example with the explicit
-/// call to `Drop::drop`, you'd get a compiler error.
-///
-/// If you'd like explicitly drop a value, [`std::mem::drop`] can be used instead.
-/// This function will take ownership of its argument and drop it. For more, see its documentation.
-///
-/// [`std::mem::drop`]: ../../std/mem/fn.drop.html
-///
-/// ## Destructors are recursive
-///
-/// If your type is something like a `struct` or `enum` that is an aggregate of other types,
-/// Rust will first call `Drop::drop` on the type, and then recursively on everything it contains.
-/// This happens even if the type does not implement `Drop` itself.
-/// For example:
+/// When a value goes out of scope, the compiler automatically inserts "drop glue"
```

Comment:
This is the first paragraph in a section titled

> ## Implementing `Drop`

I don't think that is good. A large part of the confusion (at least for me) around drop is that we have `Drop` impls and "drop glue", and they are not the same thing. So If eel there should be a section titled `## Drop Glue` that explains what drop glue is, separately (and maybe after? not sure) from the section that explains `impl Drop`.

Reply:
true

Reply:
I though about this some more and am still not completely satified.

I first considered splitting "implementing" and "using/drop glue" into 2 different sections,
but was unable to think of a way to explain one of them without requiring the other one.

I renamed `Implementing Drop` to `Using the Drop trait` for now, which makes at least some sense :thinking:

Reply:
Naively I thought explaining the trait first should work, but probably I am missing a dependency?

### Implementing `Drop`

To customize the operations that are automatically performed when a value goes out of scope, ...

### Drop glue

Whenever a value goes out of scope, the compiler automatically inserts "drop glue" which calls the `Drop` impl of that value (if it exists), and then recursively calls the "drop glue" for each field of a `struct`/`enum`.

Reply:
> ## Implementing `Drop`
>
>To customize the operations that are automatically performed when a value goes out of scope, ...

My problem with this approach is that without mentioning`DropGlue`, while reading the "Implementing `Drop`" section one can reasonably think that one has to manually call drop for the `Vec` in `struct A(Vec<u8>)`.

So while I could add an example just showing how to implement this trait without explaining how destructors work, I think that this would cause more harm than good.

Reply:
Hm, I see. In that case the order of things you have probably makes sense. But the section here now feels like it mainly explains how the `Drop` trait is called by the compiler, so shouldn't the section title be "Drop glue"?

Reply:
Don't really like "drop glue" as a section title. "Using the `Drop` trait" is not optimal, but I think it works well enough. This section shows both how `Drop` is used by the compiler and how to implement it yourself.